src/rubyfmt.rb

```diff
@@ -2477,17 +2551,27 @@ def on_string_literal(*args, &blk)
           next if part[1].nil?
           case part[1][0]
           when :@tstring_content
-            fixup_tstring_content_for_double_quotes(part[1][1])
+            fixup_tstring_content_for_double_quotes(
+              part[1][1],
+              is_single_quote: true,
+              delimiter: quote,
+            )
           else
             raise "got non tstring content in single string"
           end
         end
-      elsif [")", "}", "|"].include?(quote)
+      elsif /\W/ === quote && quote != "\""
+        next_string = @string_stack.pop
+        is_single_quote =  next_string.start_with?("%q")
         (args[0][1..-1] || []).each do |part|
           next if part.nil?
           case part[0]
           when :@tstring_content
-            fixup_tstring_content_for_double_quotes(part[1])
+            fixup_tstring_content_for_double_quotes(
+              part[1],
+              is_single_quote: is_single_quote,
+              delimiter: quote,
```

Comment:
`quote` might be the wrong thing to pass here, because in some cases the opening and closing delimiters are different (e.g. `%q(` and `)`) and in those cases we need to unescape closing delimiters, not opening ones.

We can push the closing quotes to a stack in `#on_tstring_end`. We could even use the same stack, and pop twice to get the distinct quotes:

```ruby
next_string_end = @string_stack.pop
next_string_start = @string_stack.pop
```

Reply:
can you provide a sample string that you think might surface this bug and check it?

Reply:
I thought a string like this would be a problem, but I was wrong:

```ruby
%<foo\>>
```

I was mixing up `quote` (which didn't change in this PR) with `next_string` (which is new, and came from the stack).

I think things would be more clear if we grabbed the opening and closing quotes in the same way, but there is no bug.

Reply:
I could go either way on this, how strongly do you feel I should maintain both in the stack?

Reply:
I feel more strongly about the names: `next_string` and `quote` doesn't express to me how closely related they are.

I feel less strongly about how we find the two things, but I find an `#on_tstring_end` method easier to read than `[@file_lines[lineno-1].bytes[column-1]].pack("c*")`